src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs

```diff
@@ -96,13 +98,26 @@ async Task CompleteMessageIfNeededAsync(IMessageSession session, Message message
             }
         }
 
-        async Task AbandonMessageIfNeededAsync(IMessageSession session, Message message)
+        async Task AbandonMessageIfNeededAsync(IMessageSession session, Message message, Exception callbackException)
         {
             try
             {
                 if (session.ReceiveMode == ReceiveMode.PeekLock)
                 {
-                    await session.AbandonAsync(message.SystemProperties.LockToken).ConfigureAwait(false);
+                    IDictionary<string, object> propertiesToModify = null;
+                    if (this.sessionHandlerOptions.PropertiesToModifyOnExceptionHandler != null)
+                    {
+                        var eventArgs = new ExceptionReceivedEventArgs(callbackException, ExceptionReceivedEventArgsAction.UserCallback, this.endpoint, this.entityPath, this.clientId);
+                        try
+                        {
+                            propertiesToModify = this.sessionHandlerOptions.PropertiesToModifyOnExceptionHandler(eventArgs);
+                        }
+                        catch (Exception exception)
+                        {
+                            MessagingEventSource.Log.MessageReceiverPumpUserCallbackException(this.clientId, message, exception);
```

Comment:
What this exception will look like? Will it contain enough info to understand that it's the user provided callback that has failed?

Reply:
At this point, the user callback threw an exception so the pump is abandoning the message. We let the user execute a second callback to return a dictionary of properties to modify on the message being abandoned, most likely to store details about the exception on the message itself. If that second callback fails, we don't want to fail abandoning the message so we just ignore the properties to modify and pass `null` (the default value). I'm not sure what else could be done, did you have something in mind?

Reply:
Logging that exception? My thought is the following:
Processing callback fails, message is going to be abandoned. Properties modifier callback is failing, the same log message will be made with a different exception, using `MessagingEventSource.Log.MessageReceiverPumpUserCallbackException()` variant. Perhaps it should be a more specific Log method to distinguish between callbacks? That would suffice imo.